File: matlab/example_prediction/predict3d.m
function [fixed_data,angles] = predict3d(data)

data_in=imresize3(data,[128,128,128]);

rot_table=readtable('rot_dict_unique.csv');
unique_rots=rot_table{:,1:3};

load('models/net3d.mat','net')

MEAN=900.3071;
STD=318.11615;

data_in=(data_in-MEAN)/STD;
res=predict(net,data_in);


[~,ind]=max(res);

angles=unique_rots(ind,:);



fixed_data=rotate_3d_inverse(data,angles);





end

File: python/example_prediction/predict_3d.py
import torch 
from rotate_fcns import rotate_3d,rotate_3d_inverse
import numpy as np
import pandas as pd 
from small_resnet3D import Small_resnet3D
from scipy.ndimage import zoom

def predict_3d(data,device="cuda:0"):
    
    img=data.copy()
    
    device = torch.device(device)

    model=Small_resnet3D(input_size=1,output_size=24,lvl1_size=4)
    model.load_state_dict(torch.load('models/net3d.pt')) 
    
    
    model=model.to(device)
    model=model.eval()
    
    
    
    factor=np.array([128,128,128])/img.shape
                
    img=zoom(img,factor,order=1)
            
            
            

    img=img.astype(np.float32)
    MEAN=614.2868
    STD=614.2868
    img=(img-MEAN)/STD
    
        
    
    
    
    
        
    
    img=np.expand_dims(img, axis=0)
    img=np.expand_dims(img, axis=0).copy()

    img=torch.from_numpy(img)
    img=img.to(device)
    
    res=model(img)
    res=torch.softmax(res,1)
    res=res.detach().cpu().numpy()
    pred=np.argmax(res,1)
    
    
    tmp = pd.read_csv("rot_dict_unique.csv") 
    rots_table=tmp.loc[:,:].to_numpy()
    
    
    pred_rot=rots_table[pred,:3]
    angles=pred_rot.tolist()[0]
    fixed_data=data.copy()
    fixed_data=rotate_3d_inverse(fixed_data,angles)
    
    return fixed_data,angles